bson-codec-design/DESIGN.md

```diff
@@ -0,0 +1,177 @@
+types implement the `bson.Marshaler` and `bson.Unmarshaler` interfaces. This turned out to be too
+restrictive because users sometimes want to handle the processing for types they do not define
+themselves. For example, users who generate types using a protocol buffer library
+or users who want to handle the processing of types in the standard library.
+
+The design of the v1 encoder and decoder for the Go driver presents problems when users wish to
+have more control over how specific types are handled when encoding or decoding. The control that
+users require exists on two main levels:
+- control over how individual types are encoded or decoded
+- control over how tags are processed
+
+The granularity of control users require can be defined in three levels:
+- Control over how individual types are encoded and decoded when they are embedded in another type,
+such as a map or a struct
+- Control over how individual types are encoded and decoded when they are provided directly to the
+encoder or decoder
+- Control over how tags are processed and the parameters that are returned
+
+The proposed design allows for the first two granularities to be handled together.
+
+## High Level Design
+### Registries
+To enable users to control how types are encoded and decoded, the proposed design uses a type
+registry. This registry will be provided in any context where encoding or decoding occurs. The registry
+will be used to lookup a type's codec, which will then be used for the actual encoding a decoding.
+This will allow users to control nearly all aspects of the encoding and decoding process. One
+failing of the v1 design is that users cannot control the encoding and decoding of types
+they do not own, e.g. `time.Time`. The proposed design handles this by allowing users to register a
+codec for any type.
+
+Pointer and value types are handled separately in the registry as this simplifies the registry and
+the implementations of codecs, since they will only have to worry about a single type and not the
+type plus a pointer to that type. While likely not useful, there is no design
+reason to prevent users from registering codecs for internal types like `bson.Element`,
+`bson.Document`, or `bson.Reader`. Users can override the functionality of built in types as
+well. The registry will be used for all encoding and decoding operations, since the cost of calling
+`reflect.TypeOf` is inexpensive.
+
+There is a single type, `Registry`, that contains an instance of a `TypeRegistry` and
+`InterfaceRegistry`. The `Registry` type is used to find the required codec during codec processing.
+The `Registry` type will also have a generic codec registry, which only contains two codecs, one for
+structs and one for maps. These codecs are only used if both the `TypeRegistry` and the
+`InterfaceRegistry` do not return a codec. The user can override these generic codecs.
+
+#### No Global Registries
```

Comment:
I'm not sure this is a good idea. 
1. It means everyone using Marshal and Unmarshal now have to care about something that many will never need to even know exists.
2. It means that registering the mgobson stuff automatically is impossible.

I think our internal functions should always take a registry. We as go-driver devs are required to pass one around. But our users should be required to care. If they do care, there are methods that take a registry (MarshalWithRegistry, a mongo.Client option for setting the registry, etc...)

Reply:
> It means everyone using Marshal and Unmarshal now have to care about something that many will never need to even know exists.

This is exactly the thing that the article argues against, having things that can subtly change the behavior of your application without you having any knowledge of it until it breaks. Having Marshal and Unmarshal use a registry that can change depending on the packages you import and the packages that those packages import is the problem we're trying to avoid here. This is the main problem with the `net/http` global stuff, if some package that's down in your dependency tree imports `net/http/pprof` then all of the sudden you're exposing all this internal stuff you didn't mean to. It's even worse when you try and debug where it's actually coming from, because rarely is that an easy task.

I would rather get rid of the `Marshal` and `Unmarshal` functions and provide a way to handle `[]byte` directly in the `Encoder` and `Decoder` types. I wouldn't be opposed to keeping the `Marshal` and `Unamrshal` functions and using a default registry that cannot be modified in them but that seems rather unuseful.

This stuff might seem rather trivial when you're implementing a small and straight forward application, but I've been burned enough times by this magical "packages auto register" stuff enough times that I would rather not propagate the problem further. I'm against providing users with footguns when it can be avoided.

Reply:
I think having a function with produces an independent "default" copy of the registry, and variants of the public functions/interface that take a registry and others that just use the default registry, make sense.) I agree that having a global mutable registry is bad, but having immutable defaults would be fine. 

From a very high level, every other serialization library seems to satisfy people's needs using Marshaller/Unmarshaller interfaces and doesn't require a type registry. While I think that the registry is a reasonable way to implement a decoder, we might be going about this the wrong way.  The goals of the project are: 

- have a working bson decoder that can populate a go struct, roughly on par with the functionality of the stdlib json package.
- encapsulate encoding/decoding and allow higher/lower level interfaces support bringing-your-own-bson. 
- [something else that I've missed?]

I think it might be good to consider some alternative implementations or approaches that could satisfy these higher level goals, and the relative costs of these options, which would go a long ways towards justifying the scope of this project as currently designed.